File: android/app/src/main/java/com/hcjeong/forestix/ar/BreastHeightGuide.kt
// Breast-height guide — a drawn answer to "how does the phone know it is
// reading the stem AT breast height?".
//
// The cruiser puts the crosshair on the ground at the foot of the stem and
// places a base point. A thin riser and height ring persist at its live pose;
// no timer hides the guide while the cruiser is lining up the measurement.
//
// IT IS A GUIDE AND ONLY A GUIDE. Nothing here is read by the estimator,
// reaches a QuickMeasureEntry, or appears in an export. The diameter capture
// runs byte for byte the flow it runs with the guide switched off.
//
// Twin of iOS App/BreastHeightGuide.swift: same states, same API names, same
// shapes, same grace window. Deliberately NOT a singleton like
// ArSessionHub's plot — one instance per Diameter screen, so a base placed
// at one tree's foot cannot outlive the screen that placed it.

package com.hcjeong.forestix.ar

import android.os.SystemClock
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.google.ar.core.Anchor
import com.google.ar.core.Pose
import com.google.ar.core.TrackingState
import com.hcjeong.forestix.common.UnitSystem
import com.hcjeong.forestix.common.BreastHeightGuideHeight

@Stable
class BreastHeightGuide(private val controller: ArController) {

    enum class Stage { OFF, AIMING, PLACED }

    var stage by mutableStateOf(Stage.OFF)
        private set

    var height by mutableStateOf(BreastHeightGuideHeight.METERS_130)

    /// The LIVE base point — the anchor's drift-corrected pose, re-read on
    /// every refresh, never the frozen hit coordinate the placement produced.
    /// Same rule the sampling plot follows (ArSessionHub.updatePlotNodes):
    /// an anchor translation is only meaningful in the world frame ARCore
    /// held when it was read, and that frame moves.
    var basePoint by mutableStateOf<Vec3?>(null)
        private set

    /// Where the assembly WOULD land, drawn translucent while aiming.
    var ghostPoint by mutableStateOf<Vec3?>(null)
        private set

    /// True once the pose has gone uncorrected for longer than the grace and
    /// the markers have been taken down rather than drawn in the wrong place.
    var trackingLost by mutableStateOf(false)
        private set

    private var anchor: Anchor? = null
    /// Monotonic clock, matching ArSessionHub's own choice — wall time can
    /// step and this window is half a second wide.
    private var poseStaleSinceMs = 0L

    /// Turn the guide on. Idempotent, so the Settings gate can call it on
    /// every recomposition without disturbing a base already placed.
    fun arm() {
        if (stage == Stage.OFF) stage = Stage.AIMING
    }

    /// Turn the guide off when its toggle goes off or the screen goes away.
    /// Nothing is left anchored.
    fun disable() {
        clearAnchor()
        ghostPoint = null
        trackingLost = false
        stage = Stage.OFF
    }

    /// Drop the placed base and go back to aiming — the "Reset ground" button,
    /// and the tree change in a cruise tally (the Diameter screen is reused
    /// across trees, so tree 7's foot must not still be drawn at tree 8).
    fun clearBase() {
        ghostPoint = null
        clearAnchor()
        trackingLost = false
        if (stage == Stage.PLACED) stage = Stage.AIMING
    }

    /// Anchor the base at a crosshair hit. False when the session cannot take
    /// an anchor, which is the same refusal ArSessionHub.placePlot gives.
    fun place(hit: Vec3): Boolean {
        val session = controller.session ?: return false
        val created = runCatching {
            session.createAnchor(Pose.makeTranslation(hit.x, hit.y, hit.z))
        }.getOrNull() ?: return false
        clearAnchor()
        anchor = created
        basePoint = hit
        ghostPoint = null
        trackingLost = false
        poseStaleSinceMs = 0L
        stage = Stage.PLACED
        return true
    }

    /// The ghost's live position while aiming. Null (a missed ray) draws
    /// nothing rather than freezing the preview somewhere stale.
    fun updateGhost(point: Vec3?) {
        ghostPoint = if (stage == Stage.AIMING) point else null
    }

    /// Re-read the anchor's corrected pose.
    ///
    /// The pose test is the one updatePlotNodes applies, for the reason given
    /// there: an anchor can report TRACKING for a frame or two after the
    /// camera has lost the world, so the camera's own state is part of it.
    /// A routine sub-second dip is ridden out — hiding on the first PAUSED
    /// frame makes the guide blink continuously — and past
    /// [PLOT_POSE_GRACE_MS] the guide is hidden until tracking recovers.
    fun refresh() {
        val a = anchor ?: return
        if (a.trackingState == TrackingState.STOPPED) {
            // ARCore has given up on this anchor for good; there is no pose
            // left to come back to.
            clearBase()
            return
        }
        val corrected = a.trackingState == TrackingState.TRACKING &&
            controller.frame?.camera?.trackingState == TrackingState.TRACKING
        if (corrected) {
            poseStaleSinceMs = 0L
            val p = a.pose
            basePoint = Vec3(p.tx(), p.ty(), p.tz())
            trackingLost = false
            return
        }
        val now = SystemClock.elapsedRealtime()
        if (poseStaleSinceMs == 0L) poseStaleSinceMs = now
        if (now - poseStaleSinceMs >= PLOT_POSE_GRACE_MS) trackingLost = true
    }

    /// Selected height above the live base. Ticks and label share this
    /// projection; no height is shown until a base has been anchored.
    fun heightWorldPoint(): Vec3? =
        if (stage != Stage.PLACED) null
        else drawnBase()?.let { Vec3(it.x, it.y + height.meters.toFloat(), it.z) }

    /// Label and world geometry always use the same selected height.
    fun label(system: UnitSystem): String =
        if (system == UnitSystem.METRIC) height.metricLabel else height.imperialLabel

    /// Preview dot while aiming; riser and ring persist at the tracked base.
    /// Tracking loss still hides them, but elapsed time alone never does.
    fun markers(): List<ArSceneMarker> {
        val base = drawnBase() ?: return emptyList()
        if (stage == Stage.PLACED) {
            return placementMarkers(base, height)
        }
        return listOf(ArSceneMarker(base, MarkerShape.Sphere(0.015f),
            floatArrayOf(1f, 1f, 1f, 0.45f)))
    }

    /// The point the assembly is drawn from: the anchor once placed, the aim
    /// while placing, and nothing at all once tracking has been lost long
    /// enough to distrust the pose.
    private fun drawnBase(): Vec3? = when (stage) {
        Stage.OFF -> null
        Stage.AIMING -> ghostPoint
        Stage.PLACED -> if (trackingLost) null else basePoint
    }

    private fun clearAnchor() {
        anchor?.let { runCatching { it.detach() } }
        anchor = null
        basePoint = null
        poseStaleSinceMs = 0L
    }

    companion object {
        const val PLACE_BUTTON = "Set ground"
        const val CLEAR_BUTTON = "Reset ground"

        internal fun placementMarkers(base: Vec3, height: BreastHeightGuideHeight): List<ArSceneMarker> {
            val h = height.meters.toFloat()
            val white = floatArrayOf(1f, 1f, 1f, 0.9f)
            return listOf(
                ArSceneMarker(base, MarkerShape.Sphere(0.015f), white),
                ArSceneMarker(Vec3(base.x, base.y + h / 2f, base.z),
                    MarkerShape.Cylinder(0.004f, h), white),
                ArSceneMarker(Vec3(base.x, base.y + h, base.z),
                    MarkerShape.Torus(0.35f, 0.01f), white),
            )
        }
    }
}

File: TimberCruisingApp/App/BreastHeightGuide.swift
// BREAST-HEIGHT GUIDE — the selected height above the ground, drawn in the world.
//
// A phone cannot know it is reading the stem AT breast height, and "about
// chest high" is what a cruiser is otherwise left with. This puts the height
// on screen: a persistent line from the tapped base to the selected height,
// with a thin ring at the top and a label projected from the live anchor.
//
// IT IS A GUIDE AND NOTHING ELSE. It never writes to a measurement, never
// gates the shutter, never changes a recorded diameter and never reaches an
// export — the Diameter capture runs byte-for-byte as it does with the guide
// off, in every state below.
//
// One instance per screen, deliberately NOT a singleton like
// `ActiveSamplingPlot`: the plot is a place in the stand that outlives the
// screen that placed it, whereas this base belongs to the tree in front of
// the camera right now. It dies with the screen and with the tree.
//
// It sits beside `ActiveSamplingPlot` rather than in the AR module for the
// same reason that class does: it is anchor-and-marker STATE, not rendering.
// It also has to read the plot's own `trackingGraceSeconds` — one rule about
// how long an uncorrected pose may still be drawn, in one place — and AR
// cannot see App.
//
// THE BASE IS READ, NOT PINNED, for the two reasons ActiveSamplingPlot gives
// at length: a RealityKit `AnchorEntity(.anchor(identifier:))` added to a
// scene that never saw the anchor arrive binds to nothing, and one that did
// keeps drawing at the last transform when ARKit stops tracking. A guide
// drawn at the wrong height is worse than no guide, so the pose is re-read
// from the anchor on the screen's poll and the geometry goes away when the
// read refuses for longer than the grace window.

import Foundation
import AR
import Common
import Models
import Sensors
import simd

@MainActor
public final class BreastHeightGuide: ObservableObject {

    public enum Stage: Equatable {
        /// Gate off — nothing exists. No anchor, no markers, no label.
        case off
        /// Guide on, no base placed: the cruiser is aiming at the tree base
        /// and sees a small base preview where it would land.
        case aiming
        /// Base anchored; the assembly is drawn at the anchor's live pose.
        case placed
    }

    @Published public private(set) var stage: Stage = .off
    @Published public var height: BreastHeightGuideHeight = .meters130

    /// The base point as ARKit is currently correcting it — not the frozen
    /// coordinate the placing raycast returned. nil while nothing is placed
    /// and once tracking has been lost past the grace window, and everything
    /// drawn is drawn from THIS, so nil means nothing is drawn.
    @Published public private(set) var basePoint: SIMD3<Float>?

    /// Live crosshair hit while aiming — the ghost preview's position, and
    /// what `place(hit:using:)` would anchor. nil when the ray misses, which
    /// draws nothing rather than guessing.
    @Published public private(set) var ghostPoint: SIMD3<Float>?

    /// True once the base pose has gone stale beyond the tracking grace;
    /// the guide is hidden until tracking recovers.
    @Published public private(set) var trackingLost = false

    private var anchorID: UUID?

    /// When the pose stopped being corrected, on a MONOTONIC clock.
    /// `ProcessInfo.systemUptime` and not `Date()`, for the reason
    /// `ActiveSamplingPlot.poseStaleSince` spells out: a wall clock an NTP
    /// correction can step would either hold the guide at an uncorrected pose
    /// long past the grace or blink it off while tracking was fine.
    private var poseStaleSince: TimeInterval?

    public init() {}

    // MARK: - Label

    /// Label and world geometry always use the same selected height.
    public func label(in system: UnitSystem) -> String {
        system == .metric ? height.metricLabel : height.imperialLabel
    }

    // MARK: - State

    /// Gate on: start aiming. No-op once a base is placed, so a settings
    /// re-read or a re-poll never drops the base the cruiser just planted.
    public func arm() {
        guard stage == .off else { return }
        stage = .aiming
        ghostPoint = nil
    }

    /// Gate off: forget everything and take the anchor with it. Called when
    /// the toggle goes off and on screen teardown — an
    /// anchor left behind lives in the app-shared session for the rest of the
    /// process with nobody holding its id.
    public func disable(using session: ARKitSessionManager) {
        removeAnchor(using: session)
        stage = .off
        ghostPoint = nil
        basePoint = nil
        trackingLost = false
        poseStaleSince = nil
    }

    /// Drop the base and go back to aiming — the "Reset ground" button, and the
    /// tree change in the cruise tally. The Diameter screen is reused across
    /// trees, so a base at tree 7's foot must not still be drawn at tree 8.
    /// No-op while the gate is off, which keeps the tree-change hook from
    /// arming a guide nobody asked for.
    public func clearBase(using session: ARKitSessionManager) {
        guard stage != .off else { return }
        removeAnchor(using: session)
        stage = .aiming
        ghostPoint = nil
        basePoint = nil
        trackingLost = false
        poseStaleSince = nil
    }

    /// Where the ghost goes while aiming. nil is a miss and draws nothing.
    public func updateGhost(_ hit: SIMD3<Float>?) {
        guard stage == .aiming else { return }
        ghostPoint = hit
    }

    /// Anchor the base at a hit. Returns false when the session cannot create
    /// an anchor, leaving the previous guide unchanged.
    @discardableResult
    public func place(hit: SIMD3<Float>,
                      using session: ARKitSessionManager) -> Bool {
        guard let id = session.addWorldAnchor(
            at: hit, name: "forestix.breastHeight.base")
        else {
            return false
        }
        removeAnchor(using: session)
        anchorID = id
        stage = .placed
        ghostPoint = nil
        // Where the anchor is is unknown until a poll reads it off a tracked
        // frame; nothing is drawn in between, which is one poll tick — the
        // tap that got here needed tracking anyway.
        basePoint = nil
        trackingLost = false
        poseStaleSince = nil
        return true
    }

    /// Re-read the base from its ARAnchor, or hide the guide once the pose
    /// has gone uncorrected for longer than the grace window. Inside the
    /// window the last corrected point is kept, which is what stops a routine
    /// sub-second tracking dip from blinking the whole assembly.
    ///
    /// The window is `ActiveSamplingPlot.trackingGraceSeconds` READ, not
    /// restated: it is one rule about how long an uncorrected pose may still
    /// be drawn, and a second copy of the number here would be a second rule.
    @discardableResult
    public func refresh(using session: ARKitSessionManager) -> SIMD3<Float>? {
        guard stage == .placed, let anchorID else {
            basePoint = nil
            trackingLost = false
            poseStaleSince = nil
            return nil
        }
        if let live = session.trackedWorldAnchorPosition(id: anchorID) {
            poseStaleSince = nil
            if trackingLost { trackingLost = false }
            let moved = basePoint.map {
                simd_distance($0, live) >= Self.poseEpsilonM
            } ?? true
            if moved { basePoint = live }
            return basePoint
        }
        let now = ProcessInfo.processInfo.systemUptime
        let staleSince = poseStaleSince ?? now
        poseStaleSince = staleSince
        if now - staleSince >= ActiveSamplingPlot.trackingGraceSeconds {
            if !trackingLost { trackingLost = true }
            basePoint = nil
        }
        return basePoint
    }

    private func removeAnchor(using session: ARKitSessionManager) {
        if let anchorID { session.removeWorldAnchor(id: anchorID) }
        anchorID = nil
    }

    /// Movement below this (1 mm) does not re-publish the base or rebuild
    /// marker geometry. Matches the plot's `poseEpsilonM`.
    private static let poseEpsilonM: Float = 0.001

    // MARK: - Geometry

    /// Selected height above the live base. Ticks and label share this
    /// projection and disappear together when the anchor is unavailable.
    public var heightWorldPoint: SIMD3<Float>? {
        guard stage == .placed, !trackingLost else { return nil }
        return basePoint.map { height.point(above: $0) }
    }

    /// The point the assembly is drawn from — the anchored base once placed,
    /// the live crosshair hit while aiming.
    private var drawPoint: SIMD3<Float>? {
        switch stage {
        case .off:     return nil
        case .aiming:  return ghostPoint
        case .placed:  return basePoint
        }
    }

    // Stable ids so `ARCameraView` diffs these anchors instead of rebuilding
    // their meshes on every body evaluation. Hex-only UUIDs — see the note on
    // `DBHScanScreen.cylinderMarkerId`.
    private static let baseSphereId =
        UUID(uuidString: "00B4EA17-0000-0000-0000-000000000001") ?? UUID()
    private static let riserId =
        UUID(uuidString: "00B4EA17-0000-0000-0000-000000000002") ?? UUID()
    private static let ringId =
        UUID(uuidString: "00B4EA17-0000-0000-0000-000000000003") ?? UUID()
    /// Preview dot while aiming; riser and ring stay for the lifetime of the
    /// tracked ground anchor. No timer hides an otherwise valid guide.
    public func markers() -> [ARSceneMarker] {
        guard let base = drawPoint else { return [] }
        let white = SIMD4<Float>(1, 1, 1, stage == .placed ? 0.9 : 0.45)
        var markers = [ARSceneMarker(id: Self.baseSphereId, worldPosition: base,
                                    shape: .sphere(radiusM: 0.015), colorRGBA: white)]
        if stage == .placed {
            let h = Float(height.meters)
            markers.append(ARSceneMarker(id: Self.riserId,
                worldPosition: base + SIMD3<Float>(0, h / 2, 0),
                shape: .cylinder(radiusM: 0.004, heightM: h), colorRGBA: white))
            markers.append(ARSceneMarker(id: Self.ringId,
                worldPosition: base + SIMD3<Float>(0, h, 0),
                shape: .torus(radiusM: 0.35, tubeM: 0.01), colorRGBA: white))
        }
        return markers
    }
}
